### PR TITLE
Fix MessageDef parsing of messages

### LINKED ISSUE #12393: Some problems with MessageDef parsing of messages
#### Problem
MessageDef parsing of toplevel messages expects the caller to call Next() to position the reader on the single struct in the message before initializing the MessageDef parser.  Then it checks the type, but not the tag, of that toplevel struct.  And it never checks that there's nothing after the struct.

 #### Proposed Solution
For toplevel messages, the MessageDef parser itself should do `Next(TLV::kTLVType_Structure, TLV::AnonymousTag)`, verifying the tag in the process.  And should consider checking that after the outermost struct is closed there is nothing else in the TLV (e.g. in the destructor of the parser?).

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
* Fixes #12393

#### Change overview
--For IM message, do Next(TLV::kTLVType_Structure, TLV::AnonymousTag) check when init message parser.
--Checks the type, but not the tag, of that toplevel struct for IM message's end

#### Testing
Update the unit tests.

### CODE REVIEW

File: src/app/CommandHandler.cpp [hunk 1]
```diff
@@ -147,6 +146,7 @@ CHIP_ERROR CommandHandler::ProcessInvokeRequest(System::PacketBufferHandle && pa
     {
         err = CHIP_NO_ERROR;
     }
+    ReturnErrorOnFailure(invokeRequestMessage.ExitContainer());
     return err;
```
Comment (bzbarsky-apple) [inline src/app/CommandHandler.cpp:150]: I would expect something more like:
```suggestion
    ReturnErrorOnFailure(err);
    ReturnErrorOnFailure(invokeRequestMessage.ExitContainer());
    return CHIP_NO_ERROR;
```
and similar in other places.

Comment (yunhanw-google) [reply src/app/CommandHandler.cpp]: Done，thanks

**Review by bzbarsky-apple — APPROVED** [review]

**Review by yufengwangca — APPROVED** [review]

**Comment by andy31415** [comment]
@yunhanw-google - Darwin build failed several times. Retried again because flaky, but I am unsure if real failure or not.

#15314 will disable some tests that are known flaky, so maybe that helps, however this may need more investigation. if this PR makes a flake more reproducible that may be a good thing, but still need a green somehow.

**Pull request merged**